packages/tools-iottwinmaker/src/lib/utils.ts

```diff
@@ -31,4 +32,54 @@ async function verifyWorkspaceExists(workspaceId: string) {
   console.log(`Verified workspace ${workspaceId} exists.`);
 }
 
-export { delay, verifyWorkspaceExists };
+/**
+ * Helper function for creating random IDs for S3 buckets and roles
+ * @returns random hex ID
+ */
+function createUniqueId() {
+  const uniqueId = randomBytes(4).toString('hex');
+  return uniqueId;
+}
+
+/**
+ * Helper function for filling in role and policy templates during workspace creation
+ * @param template JSON stringified role with missing values
+ * @param dict dictionary of the missing values to fill in the template
+ * @returns the complete JSON stringified template filled in
+ */
+function replaceTemplateVars(template: string, dict: Record<string, string>): string {
+  let result = template;
+  for (const key in dict) {
+    if (Object.prototype.hasOwnProperty.call(dict, key)) {
+      const value = dict[key];
+      const regex = new RegExp('{' + key + '}', 'g');
+      result = result.replace(regex, value);
+    }
+  }
+  return result;
+}
+
+/**
+ * Helper function for retrying workspace creation during role propogation
+ * @param func function to retry
+ * @param attempts number of attempts
+ * @returns generic promise for the passed in function return value
+ */
+async function retryFunction<T>(func: () => Promise<T>, attempts: number) {
```

Comment:
The documentation on this needs to be generalized or rename the function. If this is just to retry creating a workspace then it doesn't need to take a function, it can just take workspace id and try with existing credentials. If we want a generalized retryFunction method then it should not report that it is waiting for role creation when it receives a validation exception. This method should either be specific or general, but not both